README.md
# Name

Net::Telnet::Netgear - Generate and send Netgear Telnet-enable packets through Net::Telnet

# Synopsis

```perl
use Net::Telnet::Netgear;
my $telnet = Net::Telnet::Netgear->new (
    # Standard Net::Telnet parameters are allowed
    host             => 'example.com',
    packet_mac       => 'AA:BB:CC:DD:EE:FF', # or AABBCCDDEEFF
    packet_username  => 'admin',
    packet_password  => 'hunter2',
    netgear_defaults => 1
);
# The magic is done transparently: the packet has already been sent,
# if necessary, and the standard Net::Telnet API can now be used.
my @lines = $telnet->cmd ('whoami');

use Net::Telnet::Netgear::Packet;
# Manually create a packet.
my $packet = Net::Telnet::Netgear::Packet->new (mac => '...');
say length $packet->get_packet; # or whatever you want
$packet = Net::Telnet::Netgear::Packet->from_base64 ('...');
$packet = Net::Telnet::Netgear::Packet->from_string ('...');
```

# Description

This module allows to programmatically generate and send magic Telnet-enabling packets for
Netgear routers with a locked Telnet interface. The packet can either be user-provided or it can
be automatically generated given the username, password and MAC address of the router. Also, this
module is capable of sending packets using TCP or UDP (the latter is used on new firmwares), and
can automatically pick the right protocol to use, making it compatible with old and new firmwares
without any additional configuration.

The work on the Telnet protocol is done by [Net::Telnet](https://metacpan.org/pod/Net::Telnet), which is subclassed by this module.
In fact, it's possible to use the entire [Net::Telnet](https://metacpan.org/pod/Net::Telnet) API and configuration parameters.

# Methods

[Net::Telnet::Netgear](https://metacpan.org/pod/Net::Telnet::Netgear) inherits all methods from [Net::Telnet](https://metacpan.org/pod/Net::Telnet) and implements the following new
ones.

## new

```perl
my $instance = Net::Telnet::Netgear->new (%options);
```

Creates a new `Net::Telnet::Netgear` instance. Returns `undef` on failure.

`%options` can contain any of the options valid with the constructor of [Net::Telnet](https://metacpan.org/pod/Net::Telnet),
with the addition of:

- `packet_mac => 'AA:BB:CC:DD:EE:FF'`

    The MAC address of the router where the packet will be sent to. Each non-hexadecimal character
    (like colons) will be removed.

- `packet_username => 'admin'`

    The username that will be put in the packet. Defaults to `Gearguy` for compatibility reasons.
    With new firmwares, the username `admin` should be used.

    Has no effect if `packet_mac` is not specified.

- `packet_password => 'password'`

    The password that will be put in the packet. Defaults to `Geardog` for compatibility reasons.
    With new firmwares, the password of the router interface should be used.

    Has no effect if `packet_mac` is not specified.

- `packet_content => 'string'`

    The content of the packet to be sent, as a string.

    Only makes sense if the packet is not defined elsewhere.

- `packet_base64 => 'b64_string'`

    The content of the packet to be sent, as a Base64 encoded string.

    Only makes sense if the packet is not defined elsewhere.

- `packet_instance => ...`

    A subclass of [Net::Telnet::Netgear::Packet](https://metacpan.org/pod/Net::Telnet::Netgear::Packet) to be used as the packet.

    Only makes sense if the packet is not defined elsewhere.

    **NOTE:** Packets generated with ["new" in Net::Telnet::Netgear::Packet](https://metacpan.org/pod/Net::Telnet::Netgear::Packet#new),
    ["from\_string" in Net::Telnet::Netgear::Packet](https://metacpan.org/pod/Net::Telnet::Netgear::Packet#from_string) and ["from\_base64" in Net::Telnet::Netgear::Packet](https://metacpan.org/pod/Net::Telnet::Netgear::Packet#from_base64)
    can be used too.

- `packet_delay => .50`

    The amount of time, in seconds, to wait after sending the packet.
    In pseudo-code: `send_packet(); wait(packet_delay); connect()`

    Defaults to `.3` seconds, or 300 milliseconds. Can be `0`.

- `packet_wait_timeout => .75`

    The amount of time, in seconds, to wait for a response from the server before sending the packet.
    In pseudo-code: `connect(); if !can_read(in packet_wait_timeout seconds) then send_packet()`

    Only effective when the packet is sent using TCP. Defaults to `1` second.

- `packet_send_mode => 'auto|tcp|udp'`

    Determines how to send the packet. See ["packet\_send\_mode"](#packet_send_mode) below.

    Defaults to `auto`.

- `netgear_defaults => 0|1`

    If enabled, the default values defined in the hash `%Net::Telnet::Netgear::NETGEAR_DEFAULTS` are
    applied once the connection is established. See ["DEFAULT VALUES USING %NETGEAR\_DEFAULTS"](#default-values-using-netgear_defaults).

    Defaults to `0`.

- `exit_on_destroy => 0|1`

    If enabled, the `exit` shell command is sent before the object is destroyed. This is useful to
    avoid ghost processes when closing a Telnet connection without killing the shell first.

    Defaults to `0`.

## apply\_netgear\_defaults

```perl
$instance->apply_netgear_defaults;
$instance->apply_netgear_defaults (
    prompt => '/rxp/',
    cmd_remove_mode => 0
);
%Net::Telnet::Netgear::NETGEAR_DEFAULTS = (exit_on_destroy => 1);
$instance->apply_netgear_defaults;
```

Applies the values specified in the hash `%Net::Telnet::Netgear::NETGEAR_DEFAULTS`. If any
argument is specified, it is temporarily added to the hash.

See ["DEFAULT VALUES USING %NETGEAR\_DEFAULTS"](#default-values-using-netgear_defaults).

## exit\_on\_destroy

```perl
my $current_value = $instance->exit_on_destroy;
# Set exit_on_destroy to 1
my $old_value = $instance->exit_on_destroy (1);
```

Gets or sets the value of the boolean flag `exit_on_destroy`, which causes the module to send
the `exit` shell command before being destroyed. This is to avoid ghost processes when closing
a Telnet connection without killing the shell first.

## packet

```perl
my $current_value = $instance->packet;
# Set the content of the packet to '...'
my $old_value = $instance->packet ('...');
```

Gets or sets the value of the packet **as a string**. This is basically equivalent to the
`packet_content` constructor parameter.

Note that objects cannot be used - you have to call ["get\_packet" in Net::Telnet::Netgear::Packet](https://metacpan.org/pod/Net::Telnet::Netgear::Packet#get_packet)
before passing the value to this method.

## packet\_delay

```perl
my $current_value = $instance->packet_delay;
# Set packet_delay to .75 seconds
my $old_value = $instance->packet_delay (.75);
```

Gets or sets the amount of time, in seconds, to wait after sending the packet.

## packet\_send\_mode

```perl
my $current_value = $instance->packet_send_mode;
# Set packet_send_mode to 'udp'
my $old_value = $instance->packet_send_mode ('udp');
```

Gets or sets the protocol used to send the packet, between `tcp`, `udp` and `auto`.

If it is `auto`, then the module will try to guess the correct protocol to use. More specifically,
if the initial `open` performed on the specified `host` and `port` fails, the packet is sent
using UDP (and then the connection is reopened). Otherwise, if the `open` succeeds but it's
impossible to read within the ["packet\_wait\_timeout"](#packet_wait_timeout), the packet is sent using TCP.

If it is `tcp`, the packet is sent using TCP.

If it is `udp`, the packet is sent using UDP. Note that in this case the packet is always sent
before an `open` call.

**NOTE:** Generally, specifying the protocol instead of using `auto` is faster, especially when
the packet has to be sent using UDP (due to the additional connection that has to be made).

## packet\_wait\_timeout

```perl
my $current_value = $instance->packet_wait_timeout;
# Set packet_wait_timeout to 1.25
my $old_value = $instance->packet_wait_timeout (1.25);
```

Gets or sets the the amount of time, in seconds, to wait for a response from the server before
sending the packet.

Only effective when the packet is sent using TCP.

# Implementation details

When you open a connection with [Net::Telnet::Netgear](https://metacpan.org/pod/Net::Telnet::Netgear) (either with the `(fh)open` methods
inherited from [Net::Telnet](https://metacpan.org/pod/Net::Telnet) or by specifying the `host` constructor parameter), the following
actions are performed depending on the value of ["packet\_send\_mode"](#packet_send_mode).

**NOTE:** when `fhopen` is used, "socket" refers to the filehandle.

- "auto"

    This is the default. First, [Net::Telnet](https://metacpan.org/pod/Net::Telnet) tries to open the socket. If it succeeds,
    then it's assumed that the server may want a TCP packet. To check if the server actually needs
    it, a ["select" in perlfunc](https://metacpan.org/pod/perlfunc#select) call is performed on the socket to determine if data is available
    to read. If data is available, then nothing is done. Otherwise, the packet is sent using TCP and
    then the socket is re-opened.

    If the initial `open` didn't succeed, then the server is not listening on the port. It's assumed
    that the server wants an UDP packet, and it is immediately sent. The socket is re-opened, and if
    it fails again the error is propagated.

- "tcp"

    The actions specified in the first case apply, except that if the initial `open` goes wrong the
    error is immediately propagated.

- "udp"

    The packet is immediately sent before the `open` performed by [Net::Telnet](https://metacpan.org/pod/Net::Telnet). If it fails, the
    error is immediately propagated.

# Default values using %NETGEAR\_DEFAULTS

As an added feature, it's possible to enable a set of options suitable for Netgear routers.
This is possible with the hash `%Net::Telnet::Netgear::NETGEAR_DEFAULTS`, which contains a list
of methods to be called on the current instance along with their parameters. This is done by the
method ["apply\_netgear\_defaults"](#apply_netgear_defaults).

The current version specifies the following list of default values:

```perl
method              value
-----------------   -----------
cmd_remove_mode     1
exit_on_destroy     1
prompt              '/.* # $/'
waitfor             '/.* # $/'
```

It is possible to edit this list either by interacting directly with it:

```perl
$Net::Telnet::Netgear::NETGEAR_DEFAULTS{some_option} = 'some_value';
delete $Net::Telnet::Netgear::NETGEAR_DEFAULTS{some_option};
%Net::Telnet::Netgear::NETGEAR_DEFAULTS = (
    option1 => 'value1',
    option2 => 'value2'
);
```

Or you can supply additional parameters to ["apply\_netgear\_defaults"](#apply_netgear_defaults), which will be temporarily
added to the list. Note that user-specified values have priority over the ones in the hash, and
if you specify the value of an option as `undef`, it won't be set at all.

```perl
# cmd_remove_mode is set to 0 instead of 1, along with all the other
# default values
$instance->apply_netgear_defaults (cmd_remove_mode => 0);
# do not set cmd_remove_mode at all, but apply every other default
$instance->apply_netgear_defaults (cmd_remove_mode => undef);
# the standard list of default values is applied plus 'some_option'
$instance->apply_netgear_defaults (some_option => 'some_value');
# equivalent to:
{
    local %Net::Telnet::Netgear::NETGEAR_DEFAULTS = (
        %Net::Telnet::Netgear::NETGEAR_DEFAULTS,
        some_option => 'some_value'
    );
    $instance->apply_netgear_defaults;
}
```

# The magic behind timeouts

`Net::Telnet::Netgear` uses a timeout to determine if it should send the packet (using TCP).
But what's the magic behind this mysterious decimal number?

Timeouts, under normal conditions, are implemented using the ["select" in perlfunc](https://metacpan.org/pod/perlfunc#select) function (which
calls the [select(2)](http://man.he.net/man2/select) syscall). This magic function is awesome, and it works beautifully.

It would be great if the story ended here, but happy endings are pretty rare in real life.

`select` works basically everywhere when dealing with network sockets, but it doesn't work on
certain systems when dealing with generic filehandles (_Win32, I'm looking at you!_).
[Net::Telnet](https://metacpan.org/pod/Net::Telnet) can make Telnet work on arbitrary filehandles (thanks to ["fhopen" in Net::Telnet](https://metacpan.org/pod/Net::Telnet#fhopen)),
but that means that `select` may not be always available. This is a problem, and you can specify
what to do in this case with the boolean variable
`$Net::Telnet::Netgear::DIE_ON_SELECT_UNAVAILABLE`.

If this variable is false (the default), then if `select` is not available the module will simply
never send packets using TCP and emit a warning. This may not be always desiderable.

If this variable is true, then if `select` is unavailable the module will call
`Net::Telnet->error` which, when `errmode` is the default, stops the execution of the script.

**NOTE:** If ["packet\_send\_mode"](#packet_send_mode) is set to `udp`, then `select` is never called, thus
`$Net::Telnet::Netgear::DIE_ON_SELECT_UNAVAILABLE` won't have any effect even if `select` is
unavailable.

# Caveats

An `open` call may require serious amounts of time, depending on the ["packet\_send\_mode"](#packet_send_mode) and
["packet\_wait\_timeout"](#packet_wait_timeout).
Particularly, if no packet has to be sent, then `tcp` or `auto` are the fastest. Otherwise,
`udp` is the fastest (because there are no timeouts, and the packet is immediately sent).
`auto` is the slowest when the router requires the packet on UDP, because a connection is
attempted on the TCP port, while it has the same speed of `tcp` when the packet is expected on
TCP.

# See also

[Net::Telnet](https://metacpan.org/pod/Net::Telnet), [Net::Telnet::Netgear::Packet](https://metacpan.org/pod/Net::Telnet::Netgear::Packet),
[http://wiki.openwrt.org/toh/netgear/telnet.console](http://wiki.openwrt.org/toh/netgear/telnet.console),
[https://github.com/Robertof/perl-net-telnet-netgear](https://github.com/Robertof/perl-net-telnet-netgear)

# Author

Roberto Frenna (robertof DOT public AT gmail DOT com)

# Thanks

Thanks to [Derreck "insanid"](https://github.com/insanid) for the precious contribution to
the OpenWRT wiki page, and for helping me to discovery the mistery behind the "strange" packets
generated with long passwords.

Thanks to [the authors of Mojolicious](https://metacpan.org/pod/Mojolicious) for inspiration about the license and the
documentation.

# License

Copyright (C) 2014-2015, Roberto Frenna.

This program is free software, you can redistribute it and/or modify it under the terms of the
Artistic License version 2.0.


Changes
0.04 2015-06-29
  - The "Old Perl versions & Strange systems Update"
  - Replaced the operator "//" with something friendlier to old Perls.
  - Probably fixed, once and for all, the damned "packet_sending.t" test.
    There was a deadlock. Which apparently didn't show on some systems.
  - Travis-CI integration.
  - Small fixups to the code comments.

0.03 2015-03-09
  - Remove POSIX as a dependency, as it is no longer required.
  - Update the email address in the documentation.
  - Move $VERSION to a standalone variable. This fixes the module for older
    versions of Perl.
  - Support older versions of Net::Telnet.
  - Add a delay in the packet sending test, hoping to fix an unknown issue
    on some platforms.

0.02 2015-03-08
  - New option: packet_delay, which specifies the amount of time to wait after
    sending the packet. Defaults to 0.30 seconds, and can be 0.
  - Critical fix in Net::Telnet::Netgear: a missing parameter could have caused
    an infinite loop.
  - Every test is now working correctly, on every OS.
  - More tests.
  - gen_doc.pl was installed with the module. This is now fixed.
  - Makefile.PL is no longer using tabs :-)

0.01 2015-03-06
  - First release.


Makefile.PL
#!/usr/bin/env perl
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile (
    AUTHOR           => 'Roberto Frenna <robertof.public@gmail.com>',
    NAME             => 'Net::Telnet::Netgear',
    VERSION_FROM     => 'lib/Net/Telnet/Netgear.pm',
    ABSTRACT         => 'Generate and send Netgear Telnet-enable packets through Net::Telnet',
    LICENSE          => 'artistic_2',
    PREREQ_PM        => {
        map { $_ => 0 } qw(Carp Crypt::Blowfish Crypt::ECB Digest::MD5 IO::Socket::INET),
                        qw(MIME::Base64 Net::Telnet Scalar::Util)
    },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ?
       (META_MERGE => {
            resources => {
                license    => 'http://www.opensource.org/licenses/artistic-license-2.0',
                bugtracker => 'https://github.com/Robertof/perl-net-telnet-netgear/issues',
                repository => 'https://github.com/Robertof/perl-net-telnet-netgear'
            } 
        }) : ()
    )
);

# Create the target "README.md" used to generate the markdown version of the POD documentation.
# It uses the script gen_doc.pl, which requires Pod::Markdown.
# Thanks to the author of Mojolicious::Plugin::DigestAuth, who used a similar section in its
# makefile to do the same thing.
sub MY::postamble
{
    my $self = shift;
    <<END_MAKE;
README.md: $self->{VERSION_FROM}
\t\$(PERLRUN) gen_doc $self->{VERSION_FROM}
END_MAKE
}


lib/Net/Telnet/Netgear/Packet.pm
package Net::Telnet::Netgear::Packet;
use strict;
use warnings;

sub new
{
    require Net::Telnet::Netgear::Packet::Native;
    Net::Telnet::Netgear::Packet::Native->new (splice (@_, 1));
}

sub from_string
{
    require Net::Telnet::Netgear::Packet::String;
    Net::Telnet::Netgear::Packet::String->new (splice (@_, 1));
}

sub from_base64
{
    require Net::Telnet::Netgear::Packet::String;
    Net::Telnet::Netgear::Packet::String->from_base64 (splice (@_, 1));
}

sub get_packet
{
    die "Method 'get_packet' not implemented in subclass";
}

1;

=encoding utf8

=head1 NAME

Net::Telnet::Netgear::Packet - generates "telnet enable packets" for Netgear routers

=head1 SYNOPSIS

    use Net::Telnet::Netgear::Packet;
    # From a string
    my $packet = Net::Telnet::Netgear::Packet->from_string ('...');
    # From a Base64-encoded string
    my $packet = Net::Telnet::Netgear::Packet->from_base64 ('Li4u');
    # From the MAC address of the router
    my $packet = Net::Telnet::Netgear::Packet->new (
        mac      => 'AA:BB:CC:DD:EE:FF',
        username => 'admin',  # optional
        password => 'hunter2' # optional
    );
    # Gets the packet as a string.
    my $string = $packet->get_packet;

=head1 DESCRIPTION

This module allows to generate "telnet enable packets" usable with Netgear routers to unlock the
telnet interface.

You can either provide a pre-generated packet from a string or you can let the module generate
it with the MAC address of the router. It's also possible to specify the username and password
that will be put in the packet.

This module is just a wrapper - the code which handles the packets
is in C<Net::Telnet::Netgear::Packet::Native> or C<Net::Telnet::Netgear::Packet::String>,
depending on which constructor you use.

=head1 METHODS

=head2 new

    my $packet = Net::Telnet::Netgear::Packet->new (%options);

Creates a C<Net::Telnet::Netgear::Packet::Native> instance.

C<%options> can be populated with the following items:

=over 4

=item * C<< mac => 'AA:BB:CC:DD:EE' >>

The MAC address of your router. I<This is required.>

=item * C<< username => 'admin' >>

Optional, the username which will be put in the packet.
Defaults to C<Gearguy> for compatibility reasons.

=item * C<< password => 'hunter2' >>

Optional, the password which will be put in the packet.
Defaults to C<Geardog> for compatibility reasons.

=back

B<NOTE:> the packet is generated each time L</"get_packet"> is called, so it's recommended
to store the returned value in a variable and use that instead of calling the method each time.

=head2 from_string

    my $packet = Net::Telnet::Netgear::Packet->from_string ('str');

Creates a C<Net::Telnet::Netgear::Packet::String> instance.

The string has to be 128 bytes, but this check is not enforced.

=head2 from_base64

    my $packet = Net::Telnet::Netgear::Packet->from_base64 ('...');

Creates a C<Net::Telnet::Netgear::Packet::String> instance.

The decoded string has to be 128 bytes, but this check is not enforced.

=head2 get_packet

    my $packet_str = $packet->get_packet;

Retrieves the generated packet (or the user provided one).

This method must be implemented by the subclasses, and dies if it isn't (or if it is called
directly on this class).

=head1 SEE ALSO

L<Net::Telnet::Netgear>, L<http://wiki.openwrt.org/toh/netgear/telnet.console>,
L<https://github.com/Robertof/perl-net-telnet-netgear>.

=head1 AUTHOR

Roberto Frenna (robertof DOT public AT gmail DOT com)

=head1 THANKS

See L<Net::Telnet::Netgear/"THANKS">.

=head1 LICENSE

Copyright (C) 2014-2015, Roberto Frenna.

This program is free software, you can redistribute it and/or modify it under the terms of the
Artistic License version 2.0.

=cut


lib/Net/Telnet/Netgear/Packet/Native.pm
package Net::Telnet::Netgear::Packet::Native;
use strict;
use warnings;
use parent "Net::Telnet::Netgear::Packet";
use Carp;
use Crypt::ECB;
use Digest::MD5;

our @CARP_NOT = qw ( Net::Telnet::Netgear::Packet );

sub new
{
    my ($self, %opts) = @_;
    Carp::croak "Missing required parameter 'mac'"
        unless exists $opts{mac};
    $opts{mac} =~ s/\W//g;
    # WARNING: "Gearguy" and "Geardog" are no longer valid since the new firmwares of Netgear
    # routers. You need to specify the username/password combination you use to login in the
    # control panel. (default: admin/password)
    $opts{username} ||= "Gearguy";
    $opts{password} ||= "Geardog";
    Carp::croak "The MAC address and the username have to be shorter than 16 characters"
        unless length $opts{mac} < 16
        and    length $opts{username} < 16;
    # Increase the maximum password length to 33 characters.
    # See the discussion here: https://github.com/insanid/netgear-telenetenable/commit/445c972
    # Long story short: the original packet structure specifies 5 fields: username, password, mac,
    # signature, reserved. Each field excluding 'reserved' is 0x10 (16) characters. However,
    # it looks like the length of the 'password' field can be increased up to 33 characters, which
    # is the maximum length allowed for the password by the Netgear interface. The analysis of what
    # happens is in the GitHub link written before.
    Carp::croak "The password must have a maximum length of 33 characters."
        unless length $opts{password} <= 33;
    bless \%opts, $self;
}

# Generates the packet. Here's some documentation:
# http://wiki.openwrt.org/toh/netgear/telnet.console
# Special thanks to insanid@github, he's a nice guy.
sub get_packet
{
    my $self = shift;
    my ($mac, $usr, $pwd) = (
        _left_justify ($self->{mac},      0x10, "\x00"),
        _left_justify ($self->{username}, 0x10, "\x00"),
        _left_justify ($self->{password}, 0x21, "\x00")
    );
    my $text    = _left_justify ($mac . $usr . $pwd, 0x70, "\x00");
    my $payload = _swap_bytes (
        _left_justify (Digest::MD5::md5 ($text) . $text, 0x80, "\x00")
    );
    my $cipher = Crypt::ECB->new;
    $cipher->padding (PADDING_NONE);
    $cipher->cipher ("Blowfish")
        || die "Blowfish not available: ", $cipher->errstring;
    $cipher->key ("AMBIT_TELNET_ENABLE+" . $self->{password});
    _swap_bytes ($cipher->encrypt ($payload));
}

# https://goo.gl/qYA1u5
# TL;DR: The Blowfish implementation of Netgear expects data in big endian
# order, but we are using the little endian order. This fixes the problem.
# Also, from perldoc -f pack:
# - V  An unsigned long (32-bit) in "VAX" (little-endian) order.
# - N  An unsigned long (32-bit) in "network" (big-endian) order.
sub _swap_bytes
{
    pack 'V*', unpack 'N*', shift;
}

# See https://docs.python.org/2/library/string.html#string.ljust
# Left-justifies $str to $length characters, filling with $filler if necessary.
# Example:
# _left_justify 'meow', 6, 'A' -> 'meowAA'
sub _left_justify
{
    my ($str, $length, $filler) = @_;
    $filler ||= " ";
    my $i = length $str;
    return $str if $i >= $length;
    $str .= $filler while ($i++ < $length);
    $str;
}

1;


lib/Netgear/Telnet.pm
#!/usr/bin/env perl
package Netgear::Telnet 0.01;
use feature 'say';
use strict;
use warnings;
use Carp;
use IO::Select;
use IO::Socket::INET;
use Netgear::Telnet::Packet;
use Time::HiRes 'sleep';

sub new
{
    my ($self, %opts) = @_;
    croak "incorrect parameters: missing packet or/and IP\n"
        unless exists $opts{packet} and exists $opts{ip};
    croak "packet is not a subclass of Netgear::Telnet::Packet" 
        unless ref ($opts{packet}) =~ /^Netgear::Telnet::Packet/;
    # pre-generate the packet so for multiple calls we don't get slowed down
    $opts{packet} = $opts{packet}->get_packet();
    bless \%opts, $self;
}

sub send_commands
{
    my ($self, @commands) = @_;
    my $command = join "; ", map { ref $_ eq 'ARRAY' ? join " ", @$_ : $_ } @commands;
    my $rsock = IO::Socket::INET->new (
        PeerHost => $self->{ip},
        PeerPort => 23,
        Proto    => "tcp",
        Timeout  => 1
    ) || croak "can't connect to the router: $!\n";
    unless (IO::Select->new ($rsock)->can_read (1))
    {
        #say "dbg: sending packet";
        binmode $rsock;
        $rsock->send ($self->{packet});
        close $rsock;
        # unfortunately we need sleep() due to how slow
        # netgear routers are :(
        sleep (0.30);
        $rsock = IO::Socket::INET->new (
            PeerHost => $self->{ip},
            PeerPort => 23,
            Proto    => "tcp",
            Timeout  => 1
        ) || croak "can't re-establish a connection to the router: $!\n";
        croak "telnet packet is invalid, router didn't send anything on the ",
               "2nd connection" unless IO::Select->new ($rsock)->can_read (1);
    }
    # don't mind this, I was drunk
    my $delimiter =
        int (rand 512) .
        "-" .
        join (
            '',
            map
                [ '0' .. '9', 'A' .. 'Z', 'a' .. 'z' ]->[rand 62],
                1 .. 16
        )   .
        "-" .
        int (rand 512);
    say $rsock 'jtmpvar="Q"', '; echo "${jtmpvar}', $delimiter, '"; ',
        $command, "; exit";
    my ($buffer, $buffenab);
    # wait until the connection is closed
    while (<$rsock>) {
        if (!$buffenab && index ($_, "Q${delimiter}") == 0) {
            $buffenab = 1;
            next;
        }
        $buffer .= $_ if $buffenab;
    }
    close $rsock;
    $buffer =~ s/\r?\n$//gm if defined $buffer;
    return $buffer;
}

1;
__END__

=encoding utf8

=head1 NAME

Netgear::Telnet - easy-to-use telnet library designed for Netgear routers

=head1 SYNOPSIS

    use Netgear::Telnet;
    my $packet = Netgear::Telnet::Packet->new(mac => "AA:BB:CC:DD:EE");
    my $client = Netgear::Telnet->new(ip => "10.0.0.0", packet => $packet);
    # Some example commands (may not work on all routers)
    my $router_date = $client->send_commands("date");
    my $adsl_status = $client->send_commands([ "adslctl", "info" ]);
    my $my_identity = $client->send_commands("whoami", [ "echo", '$USER' ]);

=head1 DESCRIPTION

This module provides a really easy to use API to interact with the telnet
service on Netgear routers. Specifically, this library provides the
possibility to automatically enable the telnet service on a target router
by using a "telnet enable packet". The packet can either be user provided
or it can be generated automatically given the MAC address of the router.
More information is available on the OpenWRT wiki page:
L<http://wiki.openwrt.org/toh/netgear/telnet.console>

=head1 METHODS

=over 4

=item C<new( %options )>

Creates a new C<Netgear::Telnet> instance.

C<%options> must be populated with the following items:

=over 8

=item * C<< ip => '10.0.0.0' >>

The hostname or IP of the router.

=item * C<< packet => ... >>

The L<Netgear::Telnet::Packet> instance which is used to unlock the telnet
interface, if necessary. It B<won't> be used if the telnet interface is
already unlocked or if the unlock does not need to be performed.

=back

=item C<send_commands( 'command', [ 'cmd', 'with', 'params' ] )>

Opens a connection to the remote telnet server and sends the specified
commands. Any number of arguments (commands) may be provided, either
as string or as an array reference, where each item represents a parameter.

Note that B<no escaping is performed> on the commands. This means that if
you are dealing with any kind of untrusted (user-provided) data, you need
to escape it. Check out L<perlfunc/"quotemeta">.

=back

=head1 CAVEATS

This is a really basic library, it does not implement any specific thing
of the telnet protocol. You may even use it to just generate packets and
send the commands by yourself with another library.

Each time you invoke the C<send_commands> function a new connection is
made to the server (which, when the commands are sent and the data has
been received, is closed). Consequently, it's convenient to send all the
commands you need at once.

=head1 SEE ALSO

L<Info about telnet on Netgear routers in the OpenWRT
wiki|http://wiki.openwrt.org/toh/netgear/telnet.console>,
L<Netgear::Telnet::Packet documentation|Netgear::Telnet::Packet>
on how to create packets.

=head1 AUTHOR

Roberto Frenna (robertof DOT public AT gmail DOT com)

=head1 LICENSE

Copyright (c) 2014 Roberto Frenna.

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
